astropy/io/fits/hdu/base.py

```diff
@@ -133,27 +133,13 @@ class _BaseHDU(metaclass=_BaseHDUMeta):
 
     _default_name = ''
 
-    def __new__(cls, data=None, header=None, *args, **kwargs):
-        """
-        Iterates through the subclasses of _BaseHDU and uses that class's
-        match_header() method to determine which subclass to instantiate.
-
-        It's important to be aware that the class hierarchy is traversed in a
-        depth-last order.  Each match_header() should identify an HDU type as
-        uniquely as possible.  Abstract types may choose to simply return False
-        or raise NotImplementedError to be skipped.
-
-        If any unexpected exceptions are raised while evaluating
-        match_header(), the type is taken to be _CorruptedHDU.
-        """
-
-        klass = _hdu_class_from_header(cls, header)
-        return super().__new__(klass)
```

Comment:
See https://github.com/astropy/astropy/commit/a9db13e8b4b4a25a2fb94489b8604150a99243b3, this was done twice.

Reply:
This is still necessary if you want to instantiate an HDU instance directly (e.g. with `type.__call__`) and have it return the correct HDU class based on the header.  I don'e think that the existence of `_readfrom_internal` obviates its need because that's only used in alternative constructors for reading an HDU from data `_BaseHDU.fromstring` and `_BaseHDU.readfrom`.  That's different from calling `_BaseHDU.__new__` directly.

Reply:
Yeah I realized this later with  #8660, as some convenience methods relies on this. In #8666 I added a new class method to use the HDU class detection, i.e. `_BaseHDU._from_data(data, header)` instead of relying on `_BaseHDU.__call__`. I think it's a good compromise, it is no more possible to have an HDU class change is type based on the header which is a good thing imo, and we avoid the duplicated detection in the general case.